### PR TITLE
Support for rename for local filesystem

### PR DESCRIPTION
**Pull request by beroyfb**
Summary: Support for rename for local filesystem.

Differential Revision: D40912852

### CODE REVIEW

File: velox/common/file/FileSystems.cpp [hunk 1]
```diff
@@ -100,6 +100,29 @@ class LocalFileSystem : public FileSystem {
     }
   }
 
+  void rename(std::string_view path, std::string_view newPath, bool overWrite)
```
Comment (xiaoxmeng) [inline velox/common/file/FileSystems.cpp:103]: nit:
```
s/name/oldPath/
s/overWrite/overwrite or replace/
```

File: velox/common/file/FileSystems.cpp [hunk 2]
```diff
@@ -100,6 +100,29 @@ class LocalFileSystem : public FileSystem {
...
+      return;
+    }
+    int32_t rc =
+        ::rename(std::string(file).c_str(), std::string(newFile).c_str());
+    if (rc != 0) {
+      VELOX_USER_FAIL(
+          "Failed to rename file {} to {} with errno {}",
+          file,
+          newFile,
+          strerror(errno));
```
Comment (xiaoxmeng) [inline velox/common/file/FileSystems.cpp:122]: use folly::errnoStr

File: velox/common/file/FileSystems.cpp [hunk 3]
```diff
@@ -100,6 +100,29 @@ class LocalFileSystem : public FileSystem {
     }
   }
 
+  void rename(std::string_view path, std::string_view newPath, bool overWrite)
+      override {
+    auto file = extractPath(path);
```
Comment (xiaoxmeng) [inline velox/common/file/FileSystems.cpp:105]:
```
s/file/oldFile/
```

File: velox/common/file/FileSystems.cpp [hunk 4]
```diff
@@ -100,6 +100,29 @@ class LocalFileSystem : public FileSystem {
...
+    auto newFile = extractPath(newPath);
+    if (!overWrite && exists(newPath)) {
```
Comment (xiaoxmeng) [inline velox/common/file/FileSystems.cpp:107]:
```
s/overWrite/overwrite or replace/
```

File: velox/common/file/FileSystems.h [hunk 5]
```diff
@@ -49,6 +49,13 @@ class FileSystem {
   // Deletes the file at 'path'. Throws on error.
   virtual void remove(std::string_view path) = 0;
 
+  // Rename the file at 'path' to `newpath`. Throws on error.
```
Comment (xiaoxmeng) [inline velox/common/file/FileSystems.h:52]:
```
s/path/oldPath/
s/newpath/newPath/

Throws a velox user exception on error.

//
// NOTE: if 'overwrite' is true, then rename does overwrite if file at 'newPath' already exists.
```

File: velox/common/file/tests/FileTest.cpp [hunk 6]
```diff
@@ -115,6 +115,27 @@ TEST(LocalFile, viaRegistry) {
...
 
+TEST(LocalFile, rename) {
+  filesystems::registerLocalFileSystem();
+  auto tempFolder = ::exec::test::TempDirectoryPath::create();
+  auto a = fmt::format("{}/a", tempFolder->path);
+  auto b = fmt::format("{}/b", tempFolder->path);
+  auto newA = fmt::format("{}/newA", tempFolder->path);
+  auto localFs = filesystems::getFileSystem(a, nullptr);
+  {
+    auto writeFile = localFs->openFileForWrite(a);
```
Comment (xiaoxmeng) [inline velox/common/file/tests/FileTest.cpp:126]: nit:
```
localFs->openFileForWrite(fileB);
localFs->openFileForWrite(fileA);
```

File: velox/common/file/tests/FileTest.cpp [hunk 7]
```diff
@@ -115,6 +115,27 @@ TEST(LocalFile, viaRegistry) {
   lfs->remove(filename);
 }
 
+TEST(LocalFile, rename) {
+  filesystems::registerLocalFileSystem();
+  auto tempFolder = ::exec::test::TempDirectoryPath::create();
+  auto a = fmt::format("{}/a", tempFolder->path);
```
Comment (xiaoxmeng) [inline velox/common/file/tests/FileTest.cpp:121]: nit:
```
s/a/fileAPath/
s/b/fileBPath/
s/newA/newFileAPath/
```

File: velox/common/file/tests/FileTest.cpp [hunk 8]
```diff
@@ -115,6 +115,27 @@ TEST(LocalFile, viaRegistry) {
...
+    writeFile = localFs->openFileForWrite(b);
+  }
+  ASSERT_TRUE(localFs->exists(a));
+  ASSERT_TRUE(localFs->exists(b));
+  ASSERT_FALSE(localFs->exists(newA));
+  EXPECT_THROW(localFs->rename(a, b), VeloxUserError);
+  localFs->rename(a, newA);
+  ASSERT_FALSE(localFs->exists(a));
+  ASSERT_TRUE(localFs->exists(b));
+  ASSERT_TRUE(localFs->exists(newA));
```
Comment (xiaoxmeng) [inline velox/common/file/tests/FileTest.cpp:136]: Shall we add to test
```
ASSERT_TRUE(localFs->rename(b, newA, true));

Also could you write data to a and b and verify the file content after the rename? Thanks!
```

**Review by xiaoxmeng — COMMENTED** [review]
Thanks for the change!

File: velox/common/file/tests/FileTest.cpp (hunk 7 above)
Comment (beroyfb) [reply velox/common/file/tests/FileTest.cpp:121]: But there are multiple other unittests in this file that are organize that way,

Comment (beroyfb) [reply velox/common/file/tests/FileTest.cpp:121]: In terms of naming the files a or b

File: velox/common/file/tests/FileTest.cpp [hunk 9]
```diff
@@ -115,6 +115,34 @@ TEST(LocalFile, viaRegistry) {
...
+  auto tempFolder = ::exec::test::TempDirectoryPath::create();
+  auto a = fmt::format("{}/a", tempFolder->path);
+  auto b = fmt::format("{}/b", tempFolder->path);
+  auto newA = fmt::format("{}/newA", tempFolder->path);
+  auto localFs = filesystems::getFileSystem(a, nullptr);
+  {
+    auto writeFile = localFs->openFileForWrite(a);
+    writeFile = localFs->openFileForWrite(b);
+    writeFile->append("aaaaa");
+    writeFile->close();
```
Comment (xiaoxmeng) [inline velox/common/file/tests/FileTest.cpp:129]: nit: wirteFile dtor will do the close.

File: velox/common/file/tests/FileTest.cpp [hunk 10]
```diff
@@ -115,6 +115,34 @@ TEST(LocalFile, viaRegistry) {
...
+  filesystems::registerLocalFileSystem();
+  auto tempFolder = ::exec::test::TempDirectoryPath::create();
+  auto a = fmt::format("{}/a", tempFolder->path);
+  auto b = fmt::format("{}/b", tempFolder->path);
+  auto newA = fmt::format("{}/newA", tempFolder->path);
+  auto localFs = filesystems::getFileSystem(a, nullptr);
+  {
+    auto writeFile = localFs->openFileForWrite(a);
+    writeFile = localFs->openFileForWrite(b);
+    writeFile->append("aaaaa");
```
Comment (xiaoxmeng) [inline velox/common/file/tests/FileTest.cpp:128]:
```
const std::string data("aaaaa");  and use it for both write and comparison?
```

File: velox/common/file/tests/FileTest.cpp [hunk 11]
```diff
@@ -115,6 +115,34 @@ TEST(LocalFile, viaRegistry) {
...
+  ASSERT_TRUE(localFs->exists(b));
+  ASSERT_FALSE(localFs->exists(newA));
+  EXPECT_THROW(localFs->rename(a, b), VeloxUserError);
+  localFs->rename(a, newA);
+  ASSERT_FALSE(localFs->exists(a));
+  ASSERT_TRUE(localFs->exists(b));
+  ASSERT_TRUE(localFs->exists(newA));
+  localFs->rename(b, newA, true);
+  auto readFile = localFs->openFileForRead(newA);
+  char buffer1[5];
```
Comment (xiaoxmeng) [inline velox/common/file/tests/FileTest.cpp:141]:
```
char buffer[5];
```

**Review by xiaoxmeng — APPROVED** [review]
Thanks for the update. LGTM!

**Pull request closed without merging**